onnxruntime/core/providers/cpu/sequence/sequence_ops.cc

```diff
@@ -96,9 +93,228 @@ Status SequenceAt::Compute(OpKernelContext* context) const {
+ONNX_CPU_OPERATOR_KERNEL(
+    SequenceErase,
+    11,
+    KernelDefBuilder()
+        .TypeConstraint("S", DataTypeImpl::AllSequenceTensorTypes())
+        .TypeConstraint("I", std::vector<MLDataType>{
+                                 DataTypeImpl::GetTensorType<int32_t>(),
+                                 DataTypeImpl::GetTensorType<int64_t>()}),
+    SequenceErase);
+
+Status SequenceErase::Compute(OpKernelContext* context) const {
+  const auto* S = context->Input<TensorSeq>(0);
+  ORT_ENFORCE(S != nullptr, "Got nullptr for sequence input.");
+
+  const auto* I = context->Input<Tensor>(1);
+  int64_t num_tensors_input_seq = static_cast<int64_t>(S->tensors.size());
+  int64_t input_seq_idx = num_tensors_input_seq - 1;  // default is erase last one
+  if (I) {                                            // position is optional
+    input_seq_idx = GetSeqIdx(*I);
+    if (!ValidateSeqIdx(input_seq_idx, static_cast<int64_t>(num_tensors_input_seq))) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "Invalid sequence index (", input_seq_idx, ") specified for sequence of size (", num_tensors_input_seq, ")");
+    }
+
+    if (input_seq_idx < 0) {
+      input_seq_idx = static_cast<int64_t>(num_tensors_input_seq) + input_seq_idx;
+    }
+  }
+
+  auto* Y = context->Output<TensorSeq>(0);
+  ORT_ENFORCE(Y != nullptr, "SequenceErase: Got nullptr for output sequence");
+  Y->dtype = S->dtype;
+  Y->tensors.resize(num_tensors_input_seq - 1);
+  for (int i = 0, yidx = 0; i < num_tensors_input_seq; ++i, ++yidx) {
+    if (i == input_seq_idx) {
+      --yidx;
+    } else {
+      CreateAndCopyCpuTensor(S->tensors[i], context, Y->tensors[yidx]);
+    }
+  }
+
+  return Status::OK();
+}
+
+// SequenceConstruct
+ONNX_CPU_OPERATOR_KERNEL(
+    SequenceConstruct,
+    11,
+    KernelDefBuilder()
+        .TypeConstraint("T", DataTypeImpl::AllTensorTypes())
+        .TypeConstraint("S", DataTypeImpl::AllSequenceTensorTypes()),
+    SequenceConstruct);
+
+Status SequenceConstruct::Compute(OpKernelContext* context) const {
+  auto num_inputs = Node().InputArgCount().front();
+  ORT_ENFORCE(num_inputs >= 1, "Must have 1 or more inputs");
+
+  auto* Y = context->Output<TensorSeq>(0);
+  ORT_ENFORCE(Y != nullptr, "SequenceConstruct: Got nullptr for output sequence");
+
+  MLDataType first_dtype = context->Input<Tensor>(0)->DataType();
+  // Before copying check if all tensors are of the same type.
+  for (int input_idx = 0; input_idx < num_inputs; ++input_idx) {
+    const auto* X = context->Input<Tensor>(input_idx);
+    if (input_idx > 0 && X->DataType() != first_dtype) {
+      ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                      "Violation of the requirment that all input tensors must have the same data type.");
+    }
+  }
+
+  // now copy the tensors to the output sequence
+  Y->dtype = first_dtype;
+  Y->tensors.resize(num_inputs);
+  for (int input_idx = 0; input_idx < num_inputs; ++input_idx) {
```

Comment:
Technically using reserve() and push_back would be cheaper (avoiding some shared_ptr construction for the default entries that get replaced anyway) but compared to the cost of copying all the tensor data it's not going to be a significant saving.

Reply:
Sure, i can change this.